### ISSUE TITLE
FR: Revset for divergent commits

### ISSUE THREAD

**Issue by martinvonz** [jj-vcs/jj#4972]
We apparently don't have a revset function for the set of divergent commits yet. We should add that.

This came up in #4965. Filing this to track it.

**Comment by arxanas** [comment]
Related to the use-case in #4965, you could imagine wanting to clean up divergent commits by e.g. selecting whichever one has the most recent author/committer timestamp and abandoning the rest.

[currently] In the current revset language, it may be quite difficult to express that:

- We don't have a data structure in the revset language (AFAIK?) to represent a result type of "a set of sets of commits", where each set corresponds to the commits for a given divergent change ID. (However, the template language might have lists/dicts?)
- Even if we can represent the set of sets, we don't have a way to execute an operation on each set (find a representative element to preserve).
- Even if we can execute an element on each operation of the set, we don't have a way to accumulate/fold/reduce the results into a single value.

[hypothetically] You can add more control structures and data types to the revset language to model imperative computations. I suspect the most natural approach would involve adding lambdas to the language, at least.

[hypothetically] A logic programming or relational model might be able to express the above more straightforwardly than the current revset language seems to naturally support. For example, you could express this:

$$
\begin{align}
\text{commitsdiverge}(x, y) &= x \neq y \land x\text{.changeid} = y\text{.changeid} \\
\text{divergent}() &= \left\lbrace x : \text{Commit} | \exists y : \text{Commit} ~ \left[ \text{commitsdiverge}(x, y) \right] \right\rbrace \\
\text{latestdivergent}(D) &= \left\lbrace x \in D | \forall y \in D ~ \left[ \text{commitsdiverge}(x, y) \rightarrow x\text{.date} > y\text{.date} \right] \right\rbrace
\end{align}
$$

and write `jj abandon -r 'latestdivergent(divergent())'`. It's interesting to note that we've added richer "control structures" (quantifiers) to the language, without actually adding richer data structures, unlike the natural imperative extensions to the revset language. I would vaguely be in favor of switching to a logic programming paradigm, but there are significant trade-offs as well, and it's obviously that's a larger discussion that goes beyond the scope of this issue.

[action item] We should catalogue any more use-cases that aren't served well by the current revset model, which might help us decide if we want to switch paradigms for the revset language. (Unless it turns out you can handle the above case directly, in which case ignore this and let me know how 😅.)

**Comment by arxanas** [comment]
> Isn't this what jj run is?

In the current design of `jj run`:

- It accepts only a single revset, rather than a set of revsets or an alternative data structure, so it wouldn't be able to accept a set of revsets with each individual revset corresponding to a given divergent change ID.
- It can only map one input commit to one output value, without taking into consideration any other input commit, so you can't e.g. find the maximum commit in an input set according to an ordering predicate.

So the current `jj run` design doesn't provide any new functionality compared to the revset or templating languages in terms of the input or output structure. Rather, the main feature is doing the processing via arbitrary commands (possibly in the working copy, with side-effects, etc.), than just by processing the commit metadata and graph topology in-memory with the existing languages (which are limited and side-effect-free).

Conceivably, someone could design `jj run` in a way to handle the data structure and arity problems, though. I could imagine a design where `jj run` is also used in workflows for processing the commit graph in ways that are more powerful than the revset language provides, without specifically relying on side-effectful commands.

**Comment by yuja** [comment]
> `jj abandon -r 'latestdivergent(divergent())'`

Maybe I missed the point, but with the current revset language, this query might be expressed as `divergent(latest())`. Predicates like `bookmarks().filter(pat)` are usually put into the argument list.

**Comment by yuja** [comment]
I mean we can set arbitrary domain within function arguments. For example, `divergent(expr)` can be defined as a function that selects divergent commits based on `expr` where `expr` is a DSL similar to revset, but be evaluated for each divergent pair. This is basically the same as `bookmarks(pat)`. The `pat` is evaluated for each bookmark, not revision.

**Comment by arxanas** [comment]
> For example, divergent(expr) can be defined as a function that selects divergent commits based on expr where expr is a DSL similar to revset, but be evaluated for each divergent pair.

I'm skeptical that we want to invent a DSL for binary relations to represent the pairwise reduction just for `divergent`. The text pattern DSL seems more permissible because it's (presumably) applicable to many revset functions, and is a relatively simple unary predicate. But I'm open to hearing if there's an intuitive DSL to handle this case generally.

**Comment by yuja** [comment]
> I'm skeptical that we want to invent a DSL for binary relations to represent the pairwise reduction just for `divergent`.

I agree about that. It doesn't have to be a binary relation, but is still slightly different than revset expression (because the `latest(x)` revset cannot be expressed well in `divergent(latest(???))`, and `divergent(::???)` doesn't make sense.)

My feeling is that `latestdivergent(divergent())` has the same issue. `latestdivergent(D)` wouldn't be generally useful.

**Comment by arxanas** [comment]
> My feeling is that latestdivergent(divergent()) has the same issue. latestdivergent(D) wouldn't be generally useful.

Sorry, I think I was unclear in what I was proposing. I agree that `latestdivergent` isn't generally useful. Rather than proposing implementing the functions on the left-hand side of the `=`s, what I meant to propose was adding support for the logical operators on the right-hand side of all the `=`s ($\forall, \exists$, set-builder notation).

The specific example demonstrates a way that we can solve this class of problem with logic programming as a potential way to extend the language, as we don't have language features to solve the problem at present. Of course, it's possible to take many other approaches (imperative, functional, relational, etc.).

I can make a separate issue to track it. It came to mind for this issue because I thought to myself that I often want to operate on divergent commits in the way I mentioned in [my original comment](https://github.com/martinvonz/jj/issues/4972#issuecomment-2498887943), but I can't do it with the current revset language.

**Comment by scott2000** [comment]
Another option I've been thinking about recently is that we could add a `group` argument to some of the revset functions. For instance, `latest(x, group=change_id)` would first group by change ID, and then select the latest revision within each group. So now we could have `latest(divergent(), group=change_id)` to select the latest revision for each divergent change ID. This could also work for `heads()` and `roots()` I believe, and it might also make sense to group by `author` or `committer`.

This might be more natural if we had function expressions though, like `group(divergent(), change_id, |x| latest(x))`.

**Comment by yzheng453** [comment]
Hi folks, I made a [PoC implementation](https://github.com/jj-vcs/jj/compare/main...yzheng453:jj:main) to add a `divergent()` function to the revset language, similar to how `conflict()` works today.

Before moving forward I'd like to review and discuss the approach first and get some opinions.

The key question seems to be whether a simple `divergent()` is enough or more expressibility in the revset language is needed. I made my draft implementation so that it can be modified to support quantifiers like mentioned in https://github.com/jj-vcs/jj/issues/4980, but I am hesitant about moving too far in a direction that may not turn out to be favorable.

If a single `divergent()` function as it stands is useful enough I can polish and create a PR while the discussions about revset language continue.

Looking forward to inputs and thoughts.

**Comment by yuja** [comment]
I don't think we need anything special in the initial version. `divergent()` would be pretty similar to `conflicts()`. One major difference is that divergence is defined by the visible set, so we'll have to pass `visible_heads: Vec<CommitId>` to the internal `ResolvedExpression`/`ResolvedPredicateExpression`.

**Comment by yzheng453** [comment]
@yuja Thanks for the comment. Do you mind taking a look at [my implementation](https://github.com/jj-vcs/jj/compare/main...yzheng453:jj:main)? It basically takes `all()` as an argument to correctly evaluate the visibility context.

If it looks like the right direction I can add docs and tests and turn it into a PR.

**Comment by yuja** [comment]
Feel free to send a PR if you think it's ready.

> It basically takes `all()` as an argument to correctly evaluate the visibility context.

I don't think the `divergent()` revset function would take any arguments. It would be quite similar to `conflicts()` at UI level. The internal `Divergent` filter node will have to have `heads` or `visible_heads` parameter, though.

**Comment by PhilipMetzger** [comment]
This was implemented in https://github.com/jj-vcs/jj/pull/8539 which didn't contain a fixes line or an explicit link to this.
